### PR TITLE
Generalize TargetRefExists() and UpdatePolicyStatus() methods for Policy CRDs

### PR DESCRIPTION
**Pull request by zijun726911**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs from aws-gateway-controller showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this PR introduce any new dependencies?**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/utils/policy/policy_test.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	}
+	tests := []struct {
+		name                         string
+		args                         args
+		expectedK8sClientReturnedObj client.Object
+		want                         bool
+		wantErr                      error
+	}{
+		{
+			name: "AccessLogPolicy TargetRef gateway exists in the k8s, happy path",
```
Comment (xWink) [inline pkg/utils/policy/policy_test.go:37]: nit: Name should describe expected outcome, rather than just say happy path

File: pkg/utils/policy/policy_test.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+	notFoundError := fmt.Errorf("k8s resource not found")
+	type args struct {
+		policy core.Policy
+	}
+	tests := []struct {
+		name                         string
+		args                         args
+		expectedK8sClientReturnedObj client.Object
+		want                         bool
```
Comment (xWink) [inline pkg/utils/policy/policy_test.go:33]: want what?

Comment (xWink) [reply pkg/utils/policy/policy_test.go:33]: If this is just checking if we want the expected object to not be nil, shouldnt we be safe to assume that if the expectedK8sClientReturnedObj is nil, then an error is returned?

Also, expectedK8sClientReturnedObj is a bit long, can we shorten this to expectedObject?

File: pkg/utils/policy/policy_test.go (hunk 1 above)
Comment (xWink) [reply pkg/utils/policy/policy_test.go:37]: Same for all other happy path cases

**Comment by mikhail-aws** [comment]
I dont like the direction of this PR. Not clear description, not clear why moving code into different place makes it cleaner/better/easier to understand. Dont see path forward.

**Review by mikhail-aws — CHANGES_REQUESTED** [review]
Dont see improvements so far, except unit test coverage. I dont mind having better unit test coverage but please refactor this PR.

File: pkg/utils/policy/policy_test.go (hunk 2 above)
Comment (zijun726911) [reply pkg/utils/policy/policy_test.go:33]:
```
got, err := checkGatewayOrRouteTargetRefExistsForPolicy(ctx, k8sClient, tt.args.policy)
assert.Equalf(t, tt.want, got, "checkGatewayOrRouteTargetRefExistsForPolicy(%v, %v, %v)", ctx, k8sClient, tt.args.policy)
```
`want` is expected return value and `got` is actual return value, it's generated by goland idea any better name?

File: pkg/utils/policy/policy_test.go (hunk 1 above)
Comment (zijun726911) [reply pkg/utils/policy/policy_test.go:37]: Should I do "AccessLogPolicy TargetRef gateway exists in the k8s,  expect return true"? but 
```
want:    true,
wantErr: nil,
```
already indicate the expected outcome?

Comment (xWink) [reply pkg/utils/policy/policy_test.go:37]: Yes, the name of the test should be indicative of what you want from the test, regardless of params. Reason is that if a test fails, I want to read the name of the test in the output and know why it failed without having to read a bunch of code

File: pkg/utils/policy/policy_test.go (hunk 2 above)
Comment (xWink) [reply pkg/utils/policy/policy_test.go:33]: I see, it's comparing expected/actual values. If this is a common pattern, then that's fine, I was just not familiar with it

**Comment by mikhail-aws** [comment]
There might be some generic parts in two functions, it would be easier to seem them when we actually implement both of them, and common patterns emerge. Right now I dont see these common patterns in original code and this PR to make decision. These utility functions looks to specific to me, to be in generic utils.

**Comment by zijun726911** [comment]
> There might be some generic parts in two: AccessLog and Auth, it would be easier to see them when we actually implement both of them, and common patterns emerge. Right now I dont see these common patterns in original code and this PR to make decision. These utility functions looks to specific to me, to be in generic utils.

My later PR will include that code in iamauthpolicy_controller.go:

```
func (r *authPolicyReconciler) reconcileUpsert(ctx context.Context, iap *anv1alpha1.IAMAuthPolicy) error {
	if err := r.finalizerManager.AddFinalizers(ctx, iap, authPolicyFinalizer); err != nil {
		r.eventRecorder.Event(iap, corev1.EventTypeWarning,
			k8s.IAMAuthPolicyEventReasonFailedAddFinalizer, fmt.Sprintf("Failed to add finalizer due to %s", err))
		return err
	}

	if !core.AuthPolicyTargetRefExists(ctx, r.client, r.log, iap) {
		r.log.Infof("Could not find Auth Log Policy targetRef %s %s",
			iap.Spec.TargetRef.Kind, iap.Spec.TargetRef.Name)
		err := utils_policy.UpdatePolicyStatus(ctx, r.client, iap, gwv1alpha2.PolicyReasonTargetNotFound)
		if err != nil {
			return fmt.Errorf("failed to update Auth Log Policy status, %w", err)
		}
		return nil
	}

	err := r.buildAndDeployModel(ctx, iap)
	if err != nil {
		// TODO: Error handling
		return utils_policy.UpdatePolicyStatus(ctx, r.client, iap, gwv1alpha2.PolicyReasonInvalid)
	}
	err = utils_policy.UpdatePolicyStatus(ctx, r.client, iap, gwv1alpha2.PolicyReasonAccepted)
	if err != nil {
		return fmt.Errorf("failed to update Access Log Policy status, %w", err)
	}
	return nil
}
```

That use `UpdatePolicyStatus` `AuthPolicyTargetRefExists`, should I include that in this PR?
